SimG4Core/Application/src/Phase2SteppingAction.cc

```diff
@@ -0,0 +1,345 @@
+    }
+  }
+  const double time = theTrack->GetGlobalTime();
+
+  // check Z-coordinate
+  if (sAlive == tstat && std::abs(theTrack->GetPosition().z()) >= maxZCentralCMS) {
+    tstat = (time > maxTrackTimeForward) ? sOutOfTime : sVeryForward;
+  }
+
+  // check G4Region
+  if (sAlive == tstat) {
+    // next logical volume and next region
+    const G4LogicalVolume* lv = postStep->GetPhysicalVolume()->GetLogicalVolume();
+    const G4Region* theRegion = lv->GetRegion();
+
+    // kill in dead regions
+    if (isInsideDeadRegion(theRegion))
+      tstat = sDeadRegion;
+
+    // kill out of time
+    if (sAlive == tstat) {
+      if (isOutOfTimeWindow(theRegion, time))
+        tstat = sOutOfTime;
+    }
+
+    // kill low-energy in volumes on demand
+    if (sAlive == tstat && numberEkins > 0) {
+      if (isLowEnergy(lv, theTrack))
+        tstat = sLowEnergy;
+    }
+
+    // kill low-energy in vacuum
+    if (sAlive == tstat && killBeamPipe) {
+      if (theTrack->GetKineticEnergy() < theCriticalEnergyForVacuum &&
+          theTrack->GetDefinition()->GetPDGCharge() != 0.0 && lv->GetMaterial()->GetDensity() <= theCriticalDensity) {
+        tstat = sLowEnergyInVacuum;
+      }
+    }
+  }
+  // check transition tracker/btl and tracker/calo
+  bool isKilled = false;
+  if (sAlive == tstat || sVeryForward == tstat) {
+    // store TrackInformation about transition from one envelope to another
+    if (preStep->GetPhysicalVolume() == tracker && postStep->GetPhysicalVolume() == btl) {
+      // store transition tracker -> BTL only for tracks entering BTL for the first time
+      TrackInformation* trkinfo = static_cast<TrackInformation*>(theTrack->GetUserInformation());
+      if (!trkinfo->isFromTtoBTL() && !trkinfo->isFromBTLtoT()) {
+        trkinfo->setFromTtoBTL();
+        trkinfo->setIdAtBTLentrance(theTrack->GetTrackID());
+#ifdef DebugLog
+        LogDebug("SimG4CoreApplication") << "Setting flag for Tracker -> BTL " << trkinfo->isFromTtoBTL()
+                                         << " IdAtBTLentrance = " << trkinfo->idAtBTLentrance();
+#endif
+      } else {
+        trkinfo->setBTLlooper();
+        trkinfo->setIdAtBTLentrance(theTrack->GetTrackID());
+#ifdef DebugLog
+        LogDebug("SimG4CoreApplication") << "Setting flag for BTL looper " << trkinfo->isBTLlooper();
+        trkinfo->Print();
+#endif
+      }
+    } else if (preStep->GetPhysicalVolume() == btl && postStep->GetPhysicalVolume() == tracker) {
+      // store transition BTL -> tracker
+      TrackInformation* trkinfo = static_cast<TrackInformation*>(theTrack->GetUserInformation());
+      if (!trkinfo->isFromBTLtoT()) {
+        trkinfo->setFromBTLtoT();
+#ifdef DebugLog
+        LogDebug("SimG4CoreApplication") << "Setting flag for BTL -> Tracker " << trkinfo->isFromBTLtoT();
+#endif
+      }
+    } else if (preStep->GetPhysicalVolume() == tracker && postStep->GetPhysicalVolume() == calo) {
```

Comment:
@civanch are you sure this is going to work for Phase2? In my check it wasn't, as I have shown at last simulation meeting

Reply:
@fabiocos , it is a copy of SteppingAction, which does not yet used anywhere. Why I propose it ? The reason is in probably change of geometry: we need to check that a particle enter MTD, HGCal, ECAL in a different way than in Run3, because in the case of Run3 a particle go from tracker to calo directly, it is not the case for Phase-2. In the next iteration, I will add a modification in RunManager to allow identification is it Phase2 or not and selection of the corresponding SteppingAction.